TERRA-108 ⁃ Unable to import astra_access_list resource

Using Terraform 1.4.5 while trying to import an `astra_access_list`, I'm getting an error.

```
> terraform import astra_access_list.example 7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8
2023-04-26T20:06:11.897-0500 [INFO]  provider: configuring client automatic mTLS
2023-04-26T20:06:11.917-0500 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15 args=[.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15]
2023-04-26T20:06:11.920-0500 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15 pid=59456
2023-04-26T20:06:11.920-0500 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15
2023-04-26T20:06:12.317-0500 [INFO]  provider.terraform-provider-astra_v2.1.15: configuring server automatic mTLS: timestamp=2023-04-26T20:06:12.317-0500
2023-04-26T20:06:12.335-0500 [DEBUG] provider: using plugin: version=5
2023-04-26T20:06:12.335-0500 [DEBUG] provider.terraform-provider-astra_v2.1.15: plugin address: address=/var/folders/cn/r5tb0zln1bgbfzz_7x72tgzm0000gn/T/plugin3894005029 network=unix timestamp=2023-04-26T20:06:12.335-0500
2023-04-26T20:06:12.357-0500 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-04-26T20:06:12.358-0500 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15 pid=59456
2023-04-26T20:06:12.358-0500 [DEBUG] provider: plugin exited
2023-04-26T20:06:12.359-0500 [INFO]  provider: configuring client automatic mTLS
2023-04-26T20:06:12.366-0500 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15 args=[.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15]
2023-04-26T20:06:12.370-0500 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15 pid=59457
2023-04-26T20:06:12.370-0500 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15
2023-04-26T20:06:12.383-0500 [INFO]  provider.terraform-provider-astra_v2.1.15: configuring server automatic mTLS: timestamp=2023-04-26T20:06:12.383-0500
2023-04-26T20:06:12.402-0500 [DEBUG] provider.terraform-provider-astra_v2.1.15: plugin address: address=/var/folders/cn/r5tb0zln1bgbfzz_7x72tgzm0000gn/T/plugin4212217932 network=unix timestamp=2023-04-26T20:06:12.402-0500
2023-04-26T20:06:12.402-0500 [DEBUG] provider: using plugin: version=5
2023-04-26T20:06:12.426-0500 [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/datastax/astra\"]" changed the config value, but that value is unused
astra_access_list.example: Importing from ID "7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8"...
astra_access_list.example: Import prepared!
  Prepared astra_access_list for import
astra_access_list.example: Refreshing state... [id=7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8]
2023-04-26T20:06:12.428-0500 [DEBUG] provider.terraform-provider-astra_v2.1.15: 2023/04/26 20:06:12 [DEBUG] GET https://api.astra.datastax.com/v2/databases/7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8
2023-04-26T20:06:12.792-0500 [DEBUG] provider.terraform-provider-astra_v2.1.15: 2023/04/26 20:06:12 [DEBUG] GET https://api.astra.datastax.com/v2/databases/7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8/access-list
2023-04-26T20:06:12.875-0500 [ERROR] provider.terraform-provider-astra_v2.1.15: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 diagnostic_severity=ERROR diagnostic_summary="json: cannot unmarshal string into Go value of type astra.Errors" tf_proto_version=5.3 tf_rpc=ReadResource @module=sdk.proto diagnostic_detail= tf_provider_addr=provider tf_req_id=1906b29b-ec5a-3225-14d6-60ea575e31cd tf_resource_type=astra_access_list timestamp=2023-04-26T20:06:12.875-0500
2023-04-26T20:06:12.875-0500 [ERROR] vertex "import astra_access_list.example result" error: json: cannot unmarshal string into Go value of type astra.Errors
2023-04-26T20:06:12.875-0500 [ERROR] vertex "astra_access_list.example (import id \"7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8\")" error: json: cannot unmarshal string into Go value of type astra.Errors
2023-04-26T20:06:12.875-0500 [ERROR] vertex "astra_access_list.example (expand)" error: json: cannot unmarshal string into Go value of type astra.Errors
╷
│ Error: json: cannot unmarshal string into Go value of type astra.Errors
│
│
╵

2023-04-26T20:06:12.877-0500 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-04-26T20:06:12.879-0500 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/datastax/astra/2.1.15/darwin_amd64/terraform-provider-astra_v2.1.15 pid=59457
2023-04-26T20:06:12.879-0500 [DEBUG] provider: plugin exited
```

This is using a very simple terraform config file:

```
terraform {
  required_providers {
    astra = {
      source  = "datastax/astra"
      version = "~> 2.1"
    }
  }
}

resource "astra_access_list" "example" {
  database_id = "7e3b55e9-beca-43fd-b5ee-c37fa8aa16d8"
  enabled = true
  addresses {
    address = "0.0.0.1/0"
    enabled = true
  }
}
```

Seems to be very similar to #174

┆Issue is synchronized with this [Jira Task](https://datastax-oss.atlassian.net/browse/TERRA-108) by [Unito](https://www.unito.io)
┆Issue Number: TERRA-108
┆Priority: Major

--- a/internal/provider/data_source_access_list.go
+++ b/internal/provider/data_source_access_list.go
@@ -2,7 +2,10 @@ package provider
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"io"
+
 	"github.com/datastax/astra-client-go/v2/astra"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
@@ -82,28 +85,32 @@ func dataSourceAccessListRead(ctx context.Context, d *schema.ResourceData, meta 
 }
 
 func listAccessList(ctx context.Context, client *astra.ClientWithResponses, databaseID string) (*astra.AccessListResponse, error) {
-	resp, err := client.GetDatabaseWithResponse(ctx, astra.DatabaseIdParam(databaseID))
+	dbResp, err := client.GetDatabaseWithResponse(ctx, astra.DatabaseIdParam(databaseID))
 	if err != nil {
 		return nil, err
 	}
 
-	db := resp.JSON200
+	db := dbResp.JSON200
 	if db == nil {
-		return nil, fmt.Errorf("error fetching database: %s", string(resp.Body))
+		return nil, fmt.Errorf("error fetching database: %s", string(dbResp.Body))
 	}
 
-	// If the database is terminated then the private links have been deleted.
+	// If the database is terminated then the access list has been deleted.
 	if db.Status == astra.TERMINATING || db.Status == astra.TERMINATED {
 		return nil, nil
 	}
 
-	alResponse, err := client.GetAccessListForDatabaseWithResponse(ctx, astra.DatabaseIdParam(databaseID))
-
+	resp, err := client.GetAccessListForDatabase(ctx, astra.DatabaseIdParam(databaseID))
 	if err != nil {
 		return nil, err
+	} else if resp.StatusCode > 299 {
+		body, _ := io.ReadAll(resp.Body)
+		return nil, fmt.Errorf("unexpected status code: %v, message: '%s'", resp.StatusCode, body)
+	}
+	accessList := astra.AccessListResponse{}
+	if err := json.NewDecoder(resp.Body).Decode(&accessList); err != nil {
+		return nil, fmt.Errorf("failed to parse access list response: %w", err)
 	}
 
-	accessListOutput := alResponse.JSON200
-
-	return accessListOutput, err
+	return &accessList, nil
 }

--- a/internal/provider/provider.go
+++ b/internal/provider/provider.go
@@ -147,7 +147,7 @@ func configure(providerVersion string, p *schema.Provider) func(context.Context,
 		retryClient.RetryMax = 10
 		retryClient.CheckRetry = func(ctx context.Context, resp *http.Response, err error) (bool, error) {
 			// Never retry POST requests because of side effects
-			if resp.Request.Method == "POST" {
+			if resp != nil && resp.Request.Method == "POST" {
 				return false, err
 			}
 			return retryablehttp.DefaultRetryPolicy(ctx, resp, err)

--- a/internal/provider/resource_access_list.go
+++ b/internal/provider/resource_access_list.go
@@ -168,7 +168,7 @@ func resourceAccessListRead(ctx context.Context, d *schema.ResourceData, meta in
 		return diag.FromErr(err)
 	}
 
-	if string(*accessList.DatabaseId) == databaseID {
+	if accessList != nil && string(*accessList.DatabaseId) == databaseID {
 		if err := setAccessListData(d, accessList); err != nil {
 			return diag.FromErr(err)
 		}

--- a/test/run_tests.sh
+++ b/test/run_tests.sh
@@ -13,6 +13,7 @@ setup_env() {
     TEST_ENV_FILE="${SCRIPT_PATH}/${DEFAULT_TEST_ENV_FILE}"
   fi
   if [ -f "$TEST_ENV_FILE" ]; then
+    echo "loading config from file $TEST_ENV_FILE"
     source "$TEST_ENV_FILE"
   else
     echo "file '$TEST_ENV_FILE' not found, some tests may be skipped"

